### PR TITLE
FIX: do not show PM topics when moving posts to an existing public topic

### PR DESCRIPTION
**Pull request by techAPJ**
https://meta.discourse.org/t/trying-to-move-post-to-existing-pm-results-in-discourse-invalidparameters/106117/6?u=techapj

### CODE REVIEW

File: app/assets/javascripts/discourse/components/choose-topic.js.es6 [hunk 1]
```diff
@@ -32,20 +32,22 @@ export default Ember.Component.extend({
...
-              .mapBy("topic")
-              .filter(t => t.get("id") !== currentTopicId)
-          );
-        } else {
-          self.setProperties({ topics: null, loading: false });
-        }
+    searchForTerm(title, {
+      typeFilter: "topic",
+      searchForId: true,
+      restrict_to_archetype: "regular"
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/components/choose-topic.js.es6]: I don't think `searchForTerm` will process the `restrict_to_archetype` key in this object. Also, we're mixing snake_case with camelCase here which can be quite confusing.

File: lib/search.rb [hunk 2]
```diff
@@ -629,7 +629,12 @@ def add_more_topics_if_expected
 
   # If we're searching for a single topic
   def single_topic(id)
-    post = Post.find_by(topic_id: id, post_number: 1)
+    if @opts[:restrict_to_archetype].present?
+      topic = Topic.find_by(id: id, archetype: @opts[:restrict_to_archetype] == "regular" ? Archetype.default : Archetype.private_message)
```
Comment (tgxworld) [inline lib/search.rb]:
```suggestion
      topic = Topic.find_by(id: id, archetype: @opts[:restrict_to_archetype] == Archetype.default ? Archetype.default : Archetype.private_message)
```

File: spec/components/search_spec.rb [hunk 3]
```diff
@@ -341,12 +341,22 @@ def new_post(raw, topic = nil)
...
         expect(result.posts.length).to eq(1)
         expect(result.posts.first.id).to eq(post.id)
       end
+
+      context 'restrict_to_archetype' do
+        let(:personal_message) { Fabricate(:private_message_topic) }
+        let!(:p1) { Fabricate(:post, topic: personal_message, post_number: 1) }
+
+        it 'restricts result to specified archetype' do
+          result = Search.execute(personal_message.relative_url, search_for_id: true, type_filter: 'topic', guardian: Guardian.new(Fabricate(:admin)), restrict_to_archetype: 'regular')
```
Comment (tgxworld) [inline spec/components/search_spec.rb]:
```suggestion
          result = Search.execute(personal_message.relative_url, search_for_id: true, type_filter: 'topic', guardian: Guardian.new(Fabricate(:admin)), restrict_to_archetype: ArcheType.default)
```

File: spec/components/search_spec.rb [hunk 4]
```diff
@@ -341,12 +341,22 @@ def new_post(raw, topic = nil)
...
+        result = Search.execute(topic.relative_url, search_for_id: true, type_filter: 'topic')
         expect(result.posts.length).to eq(1)
         expect(result.posts.first.id).to eq(post.id)
       end
+
+      context 'restrict_to_archetype' do
+        let(:personal_message) { Fabricate(:private_message_topic) }
+        let!(:p1) { Fabricate(:post, topic: personal_message, post_number: 1) }
+
+        it 'restricts result to specified archetype' do
```
Comment (tgxworld) [inline spec/components/search_spec.rb]: We need to add a test that shows searching for a PM when restricting the archetype to "private_message" works

File: lib/search.rb [hunk 5]
```diff
@@ -629,7 +629,12 @@ def add_more_topics_if_expected
...
+      post = topic&.first_post
```
Comment (tgxworld) [inline lib/search.rb]: This ends up executing two queries as opposed to one. I think we can do a joins on the topics table to look for the post we need.

File: spec/components/search_spec.rb (hunk 3 above)
Comment (techAPJ) [reply spec/components/search_spec.rb]: Made the change in latest commit. Notice `ArcheType` vs `Archetype` :)

**Comment by techAPJ** [comment]
@tgxworld Thanks for review. Updated the PR.

**Pull request merged**